modules/460-log-shipper/hooks/internal/vector/transform/throttle.go

```diff
@@ -17,24 +17,90 @@ limitations under the License.
 package transform
 
 import (
+	"fmt"
+	"strings"
+
 	"github.com/deckhouse/deckhouse/go_lib/set"
+	"github.com/deckhouse/deckhouse/modules/460-log-shipper/apis"
 	"github.com/deckhouse/deckhouse/modules/460-log-shipper/apis/v1alpha1"
+	"github.com/deckhouse/deckhouse/modules/460-log-shipper/hooks/internal/vrl"
+)
+
+const (
+	excludeField  = "exclude"
+	keyFieldField = "key_field"
 )
 
 // ThrottleTransform adds throttling to event's flow.
-func ThrottleTransform(rl v1alpha1.RateLimitSpec) *DynamicTransform {
-	throttleTransform := DynamicTransform{
+func ThrottleTransform(rl v1alpha1.RateLimitSpec) (apis.LogTransform, error) {
+	throttleTransform := &DynamicTransform{
 		CommonTransform: CommonTransform{
 			Name:   "ratelimit",
 			Type:   "throttle",
 			Inputs: set.New(),
 		},
 		DynamicArgsMap: map[string]interface{}{
-			"exclude":     "null",
+			excludeField:  "null",
 			"threshold":   *rl.LinesPerMinute,
 			"window_secs": 60,
 		},
 	}
 
-	return &throttleTransform
+	if rl.KeyField != "" {
+		throttleTransform.DynamicArgsMap[keyFieldField] = rl.KeyField
+	}
+
+	if rl.Excludes != nil {
+		excludeCond, err := processExcludesForDynamicTransform(rl.Excludes, *rl.LinesPerMinute, rl.KeyField)
+		if err != nil {
+			return nil, err
+		}
+		throttleTransform.DynamicArgsMap[excludeField] = excludeCond
+	}
+	return throttleTransform, nil
+}
+
+func processExcludesForDynamicTransform(excludes []v1alpha1.Filter, threshold int32, keyField string) (map[string]interface{}, error) {
+	throttleTransformExcludes := make([]string, len(excludes))
+	for i, filter := range excludes {
+		condition, err := excludeThrottleTransformCond(&filter)
+		if err != nil {
+			return nil, err
+		}
+
+		throttleTransformExcludes[i] = *condition
+	}
+	resultCond := combineThrottleTransformExcludes(throttleTransformExcludes)
+	return map[string]interface{}{"type": "vrl", "source": resultCond}, nil
+}
+
+func excludeThrottleTransformCond(exclude *v1alpha1.Filter) (*string, error) {
+	if exclude == nil {
+		return nil, fmt.Errorf("no filter provided for dynamic transform exclude")
+	}
+
+	rule := getRuleOutOfFilter(exclude)
+	if rule == "" {
+		return nil, fmt.Errorf("exclude rule value is empty for dynamic transform")
+	}
+
+	condition, err := rule.Render(vrl.Args{"filter": exclude})
+	if err != nil {
+		return nil, fmt.Errorf("error rendering exclude rule for dynamic transform: %w", err)
+	}
+
+	return &condition, nil
+}
+
+func combineThrottleTransformExcludes(excludes []string) string {
+	resultExcludeConds := make([]string, len(excludes)+1)
+	resultExcludeConds[0] = "matchedExcludeCond"
+	resultCond := fmt.Sprintf("%s = false;\n", resultExcludeConds[0])
+
+	for i, excludeCond := range excludes {
+		resultExcludeConds[i+1] = fmt.Sprintf("matchedExcludeCond%d", i)
```

Comment:
I don't think you need this variable.  In VRL, each block is variable, so you can just merge them.

```rust
{ true } || { false }

true
```



Instead of rendering all the rules and saving them to an array, the much cleaner way to do this merge is using a string builder.
```go
var trottleResult strings.Builder
trottleResult.WriteString("false")

for _, filter := range excludes {
  	rule := getRuleOutOfFilter(exclude)
	condition, err := rule.Render(vrl.Args{"filter": filter})
	if err != nil {
		return nil, fmt.Errorf("error rendering exclude rule for dynamic transform: %w", err)
	}
     
    trottleResult.WriteString(" || ")
    trottleResult.WriteString(condition)
}

return trottleResult.String()
```

Reply:
It seems like an abstraction leak, vrl building in transform package, but it can live like this for a while.